### PR TITLE
Draft: Interaction Code

### PR DESCRIPTION
**Pull request by bellatumaneng-okta**
## Description:
- First draft: updates to the /token endpoint and client registration reference docs for the Interaction Code Grant.
- This draft is intentionally sparse (no links are available yet) and only covers a subset of the required updates. The goal is to get a head start on the doc updates + collect initial feedback.

### Resolves:

* [OKTA-353226](https://oktainc.atlassian.net/browse/OKTA-353226)

### CODE REVIEW

File: packages/@okta/vuepress-site/docs/reference/api/oauth-clients/index.md [hunk 1]
```diff
@@ -753,7 +753,7 @@ Client Applications have the following properties:
...
 | client_secret                           | OAuth 2.0 client secret string (used for confidential clients)                                                               | String                                                                                         | TRUE       | TRUE     | TRUE      |
 | client_secret_expires_at                | Time at which the client_secret will expire or 0 if it will not expire(measured in unix seconds)                             | Number                                                                                         | TRUE       | FALSE    | TRUE      |
-| grant_types                             | Array of OAuth 2.0 grant type strings. Default value: `authorization_code`                    | Array of `authorization_code`, `implicit`, `password`, `refresh_token`, `client_credentials`   | TRUE       | FALSE    | FALSE     |
+| grant_types                             | Array of OAuth 2.0 grant type strings. Default value: `authorization_code`                    | Array of `authorization_code`, `implicit`, `interaction_code`, `password`, `refresh_token`, `client_credentials`   | TRUE       | FALSE    | FALSE     |
```
Comment (VadymYaremchuk-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oauth-clients/index.md:756]: does it mean that interaction code grant is enabled by default?

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md [hunk 2]
```diff
@@ -271,14 +271,21 @@ The following parameters can be posted as a part of the URL-encoded form values
...
-| code_verifier           | Required if `grant_type` is `authorization_code`  and `code_challenge` was specified in the original `/authorize` request. This value is the code verifier for [PKCE](#parameter-details). Okta uses it to recompute the `code_challenge` and verify if it matches the original `code_challenge` in the authorization request.     | String |
-| grant_type              | Can be one of the following: `authorization_code`, `password`, `client_credentials`, or `refresh_token`. Determines the mechanism Okta uses to authorize the creation of the tokens.                                                                                                                                               | String |
+| code_verifier           | Required if `grant_type` is `interaction_code`, or if the `grant_type` is `authorization_code`  and `code_challenge` was specified in the original `/authorize` request. This value is the code verifier for [PKCE](#parameter-details). Okta uses it to recompute the `code_challenge` and verify if it matches the original `code_challenge` in the authorization request.     | String |
```
Comment (VadymYaremchuk-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:274]: for interaction code we have original request not to `/authorize` but is either `/token` or `/interact`

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md [hunk 3]
```diff
@@ -271,14 +271,21 @@ The following parameters can be posted as a part of the URL-encoded form values
...
 | username                | Required if the grant_type is `password`.                                                                                                                                                                                                                                                                                          | String |
 
+#### Parameter details
+* `interaction_code` grant type is an Okta extension to [the OpenID specification](http://openid.net/specs/openid-connect-core-1_0.html#Authentication).
+    All other parameters comply with the OpenID Connect specification and their behavior is consistent with the specification.
+
+* To obtain an interaction handle using the `password` grant type, the following parameters are required in addition to those specified above: `code_challenge`, `code_challenge_method`, `redirect_uri`, and `state`. See the `/interact` endpoint for more details.
```
Comment (VadymYaremchuk-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:287]: where to find `/interact` endpoint with more details?

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md [hunk 4]
```diff
@@ -291,11 +298,13 @@ Based on the scopes requested. Generally speaking, the scopes specified in a req
...
 | refresh_token   | An opaque refresh token. This is returned if the `offline_access` scope is granted.     | String  |
 | id_token        | An [ID token](#id-token). This is returned if the `openid` scope is granted.            | String  |
+| interaction_handle | An opaque handle that can be used to start a transaction with the Okta Identity Engine. See guide for more details.            | String  |
 
 #### List of errors
 
 | Error Id                 | Details                                                                                                                                                                                                    |
 | :----------------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| interaction_required     | The client is configured to use the Interaction Code Flow and user interaction is required to complete the request.                                                                                        |
```
Comment (VadymYaremchuk-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:307]: nit: may be we can say here that with this error there will be `interaction_handle` returned that need to be used for user interaction.

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md (hunk 3 above)
Comment (bellatumaneng-okta) [reply packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:287]: Adding doc for that in the next ticket. The `/interact` endpoint will be documented on the same page, and will update the link here once that's done.

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md (hunk 4 above)
Comment (bellatumaneng-okta) [reply packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:307]: Good point. @susanharper-okta can you help me with the phrasing? 

The client is configured to use the Interaction Code Flow and user interaction is required to complete the request. This error returns an `interaction_handle`, which can be used to...

There should be a link to wherever we describe the `interaction_handle` in detail. I'm trying to decide if we also need to give a quick explanation here of how the handle can be used (e.g. if I should finish that trailed off sentence or not).

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md (hunk 2 above)
Comment (bellatumaneng-okta) [reply packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:274]: Required if `grant_type` is `interaction_code`, or if the `grant_type` is `authorization_code` and `code_challenge` was specified in the original `/authorize` request. This value is the code verifier for [PKCE](#parameter-details). Okta uses it to recompute the `code_challenge` and verify if it matches the original `code_challenge` in the authorization, interact, or initial token request.

This description is getting a little long, so I'm tempted to move some of the details to the Parameter details section. @susanharper-okta thoughts?

File: packages/@okta/vuepress-site/docs/reference/api/apps/index.md [hunk 5]
```diff
@@ -1313,7 +1313,7 @@ Adds an OAuth 2.0 client application. This application is only available to the
...
-| grant_types                                 | Array of OAuth 2.0 grant type strings                                                                                                                                                                                      | Array of `authorization_code`, `implicit`, `password`, `refresh_token`, `client_credentials`   | FALSE      | FALSE    | TRUE       |
+| grant_types                                 | Array of OAuth 2.0 grant type strings                                                                                                                                                                                      | Array of `authorization_code`, `implicit`, `interaction_code`, `password`, `refresh_token`, `client_credentials` | FALSE      | FALSE    | TRUE       |
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/apps/index.md:1316]:
```suggestion
| grant_types                                 | Array of OAuth 2.0 grant type strings                                                                                                                                                                                      | Array of `authorization_code`, `client_credentials`, `implicit`, `interaction_code`, `password`, `refresh_token`  | FALSE      | FALSE    | TRUE       |
```

might as well put them in alpha order?

File: packages/@okta/vuepress-site/docs/reference/api/authorization-servers/index.md [hunk 6]
```diff
@@ -1763,7 +1763,7 @@ Example from a Policy object
...
 | clients      | For Policies, specifies which clients are included or excluded in the Policy                                                                                                                                                             | `include` and `exclude` lists   | True                          |
-| grantTypes   | Array of grantTypes that this condition includes. Accepted grantTypes: `authorization_code`, `password`, `refresh_token` or `client_credentials`. Determines the mechanism Okta uses to authorize the creation of the tokens.           | `include` list                  | True                          |
+| grantTypes   | Array of grantTypes that this condition includes. Accepted grantTypes: `authorization_code`, `interaction_code`, `password`, `refresh_token` or `client_credentials`. Determines the mechanism Okta uses to authorize the creation of the tokens. | `include` list         | True                          |
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/authorization-servers/index.md:1766]:
```suggestion
| grantTypes   | Array of grantTypes that this condition includes. Accepted grantTypes: `authorization_code`, `client_credentials`, `interaction_code`, `password`, or `refresh_token`. Determines the mechanism that Okta uses to authorize token creation. | `include` list         | True                          |
```

File: packages/@okta/vuepress-site/docs/reference/api/oauth-clients/index.md (hunk 1 above)
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oauth-clients/index.md:756]:
```suggestion
| grant_types                             | Array of OAuth 2.0 grant type strings. Default value: `authorization_code`                    | Array of `authorization_code`, `client_credentials`, `implicit`, `interaction_code`, `password`, `refresh_token`  | TRUE       | FALSE    | FALSE     |
```

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md (hunk 2 above)
Comment (susanharper-okta) [reply packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:274]: I think it's ok. Not any less meaty than the code_verifier description. And I think meaty is good. :)

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md [hunk 7]
```diff
@@ -271,14 +271,21 @@ The following parameters can be posted as a part of the URL-encoded form values
...
-| grant_type              | Can be one of the following: `authorization_code`, `password`, `client_credentials`, or `refresh_token`. Determines the mechanism Okta uses to authorize the creation of the tokens.                                                                                                                                               | String |
+| code_verifier           | Required if `grant_type` is `interaction_code`, or if the `grant_type` is `authorization_code`  and `code_challenge` was specified in the original `/authorize` request. This value is the code verifier for [PKCE](#parameter-details). Okta uses it to recompute the `code_challenge` and verify if it matches the original `code_challenge` in the authorization request.     | String |
+| grant_type              | Can be one of the following: `authorization_code`, `interaction_code`, `password`, `client_credentials`, or `refresh_token`. Determines the mechanism Okta uses to authorize the creation of the tokens.                                                                                                                           | String |
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:275]:
```suggestion
| grant_type              | Can be one of the following: `authorization_code`, `client_credentials`, `interaction_code`, `password`, or `refresh_token`. Determines the mechanism that Okta uses to authorize token creation.                                                                                                                           | String |
```

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md [hunk 8]
```diff
@@ -271,14 +271,21 @@ The following parameters can be posted as a part of the URL-encoded form values
...
 | scope                   | Required if `password` is the `grant_type`. This is a list of scopes that the client wants to be included in the access token. For the `refresh_token` grant type, these scopes have to be a subset of the scopes used to generate the refresh token in the first place.                                                             | String |
 | username                | Required if the grant_type is `password`.                                                                                                                                                                                                                                                                                          | String |
 
+#### Parameter details
+* `interaction_code` grant type is an Okta extension to [the OpenID specification](http://openid.net/specs/openid-connect-core-1_0.html#Authentication).
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:284]:
```suggestion
* `interaction_code` grant type is an Okta extension to the [OpenID specification](http://openid.net/specs/openid-connect-core-1_0.html#Authentication).
```

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md [hunk 9]
```diff
@@ -291,16 +298,18 @@ Based on the scopes requested. Generally speaking, the scopes specified in a req
 | scope           | The scopes contained in the access token.                                               | String  |
 | refresh_token   | An opaque refresh token. This is returned if the `offline_access` scope is granted.     | String  |
 | id_token        | An [ID token](#id-token). This is returned if the `openid` scope is granted.            | String  |
+| interaction_handle | An opaque handle that can be used to start a transaction with the Okta Identity Engine. See guide for more details.            | String  |
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:301]:
```suggestion
| interaction_handle | An opaque handle that you can use to start a transaction with the Okta Identity Engine. See guide for more details.            | String  |
```

Adding note here to not forget the link to the guide

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md [hunk 10]
```diff
@@ -291,16 +298,18 @@ Based on the scopes requested. Generally speaking, the scopes specified in a req
...
 
 #### List of errors
 
 | Error Id                 | Details                                                                                                                                                                                                    |
 | :----------------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| interaction_required     | The client is configured to use the Interaction Code Flow and user interaction is required to complete the request.                                                                                        |
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:307]:
```suggestion
| interaction_required     | The client is configured to use the Interaction Code flow and user interaction is required to complete the request.                                                                                        |
```

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md (hunk 4 above)
Comment (susanharper-okta) [reply packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:307]: The client is configured to use the Interaction Code flow and user interaction is required to complete the request. This error returns an `interaction_handle` that you can use to (I would def give a brief explanation here and then link off to the interaction_handle in detail content).

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md [hunk 11]
```diff
@@ -291,16 +298,18 @@ Based on the scopes requested. Generally speaking, the scopes specified in a req
...
 | invalid_grant            | The `code`, `refresh_token`, or `username` and `password` combination is invalid, or the `redirect_uri` doesn't match the one used in the authorization request.                                          |
 | invalid_request          | The request structure was invalid. For example: the basic authentication header is malformed, both header and form parameters were used for authentication, or no authentication information was provided. |
 | invalid_scope            | The scopes list contains an invalid or unsupported value.                                                                                                                                                  |
-| unsupported_grant_type   | The `grant_type` isn't `authorization_code`, `refresh_token`, or `password`.                                                                                                                               |
+| unsupported_grant_type   | The `grant_type` isn't `authorization_code`, `interaction_code`, `refresh_token`, or `password`.                                                                                                           |
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:312]:
```suggestion
| unsupported_grant_type   | The `grant_type` isn't `authorization_code`, `interaction_code`, `password`, or `refresh_token` |
```

**Review by susanharper-okta — COMMENTED** [review]
couple things.

File: packages/@okta/vuepress-site/docs/reference/api/apps/index.md [hunk 12]
```diff
@@ -1339,13 +1339,15 @@ Adds an OAuth 2.0 client application. This application is only available to the
...
 | `service`         | `client_credentials`                                          | Works with OAuth 2.0 flow (not OpenID Connect)                                    |
-| `web`             | `authorization_code`, `implicit`, `refresh_token`             | Must have at least `authorization_code`                                           |
+| `web`             | `authorization_code`, `implicit`, `interaction_code`, `refresh_token` | Must have at least `authorization_code`                                   |
 
 * The `grant_types` and `response_types` values described above are partially orthogonal, as they refer to arguments passed to different endpoints in the [OAuth 2.0 protocol](https://tools.ietf.org/html/rfc6749). However, they are related in that the `grant_types` available to a client influence the `response_types` that the client is allowed to use, and vice versa. For instance, a `grant_types` value that includes `authorization_code` implies a `response_types` value that includes `code`, as both values are defined as part of the OAuth 2.0 authorization code grant.
 
+* The `interaction_code` grant type is only available as part of the Okta Identity Engine.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/apps/index.md:1349]:
```suggestion
* The `interaction_code` grant type is only available as part of Identity engine.
```

File: packages/@okta/vuepress-site/docs/reference/api/oauth-clients/index.md [hunk 13]
```diff
@@ -784,13 +784,15 @@ Property details
...
-| `native`          | `authorization_code`, `implicit`, `password`, `refresh_token`              | Must have at least `authorization_code`        |
+| `browser`         | `authorization_code`, `implicit`, `interaction_code`                       |                                                |
+| `native`          | `authorization_code`, `implicit`, `interaction_code`, `password`, `refresh_token` | Must have at least `authorization_code`        |
 | `service`         | `client_credentials`                                                       | Works with OAuth 2.0 flow (not OpenID Connect) |
-| `web`             | `authorization_code`, `implicit`, `refresh_token`, `client_credentials`(*) | Must have at least `authorization_code`        |
+| `web`             | `authorization_code`, `implicit`, `interaction_code`, `refresh_token`, `client_credentials`(*) | Must have at least `authorization_code`        |
 
 * `client_credentials` with a `web` Application type allows you to use one `client_id` for an Application that needs to make user-specific calls and back-end calls for data.
 
+* The `interaction_code` grant type is only available as part of the Okta Identity Engine.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oauth-clients/index.md:794]:
```suggestion
* The `interaction_code` grant type is only available as part of Identity engine.
```

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md [hunk 14]
```diff
@@ -258,7 +258,7 @@ https://www.example.com/#error=invalid_scope&error_description=The+requested+sco
 
 <ApiOperation method="post" url="${baseUrl}/v1/token" />
 
-This endpoint returns access tokens, ID tokens, and refresh tokens, depending on the request parameters. For [password](/docs/guides/implement-password/), [client credentials](/docs/guides/implement-client-creds/), and [refresh token](/docs/guides/refresh-tokens/) flows, calling `/token` is the only step of the flow. For the [authorization code](/docs/guides/implement-auth-code/) flow, calling `/token` is the second step of the flow.
+This endpoint returns access tokens, ID tokens, and refresh tokens, depending on the request parameters. For [password](/docs/guides/implement-password/), [client credentials](/docs/guides/implement-client-creds/), and [refresh token](/docs/guides/refresh-tokens/) flows, calling `/token` is the only step of the flow. For the [authorization code](/docs/guides/implement-auth-code/) and interaction_code flows, calling `/token` is the last step of the flow.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:261]:
```suggestion
This endpoint returns access tokens, ID tokens, and refresh tokens depending on the request parameters. For [password](/docs/guides/implement-password/), [client credentials](/docs/guides/implement-client-creds/), and [refresh token](/docs/guides/refresh-tokens/) flows, calling `/token` is the only step of the flow. For the [authorization code](/docs/guides/implement-auth-code/) and interaction_code flows, calling `/token` is the last step of the flow.
```

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md [hunk 15]
```diff
@@ -291,16 +298,18 @@ Based on the scopes requested. Generally speaking, the scopes specified in a req
 | scope           | The scopes contained in the access token.                                               | String  |
 | refresh_token   | An opaque refresh token. This is returned if the `offline_access` scope is granted.     | String  |
 | id_token        | An [ID token](#id-token). This is returned if the `openid` scope is granted.            | String  |
+| interaction_handle | An opaque handle that you can use to start a transaction with the Okta Identity Engine. See guide for more details. | String  |
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:301]:
```suggestion
| interaction_handle | An opaque handle that you can use to start a transaction with Identity engine. See guide for more details. | String  |
```

File: packages/@okta/vuepress-site/docs/reference/api/oidc/index.md [hunk 16]
```diff
@@ -291,16 +298,18 @@ Based on the scopes requested. Generally speaking, the scopes specified in a req
...
 
 #### List of errors
 
 | Error Id                 | Details                                                                                                                                                                                                    |
 | :----------------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| interaction_required     | The client is configured to use the Interaction Code flow and user interaction is required to complete the request. This error returns an `interaction_handle` that you can use to start a transaction with the Okta Identity Engine. |
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/reference/api/oidc/index.md:307]:
```suggestion
| interaction_required     | The client is configured to use the Interaction Code flow and user interaction is required to complete the request. This error returns an `interaction_handle` that you can use to start a transaction with Identity engine. |
```

**Pull request closed without merging**